### PR TITLE
Cars

### PR DESCRIPTION
**Pull request by majkel84**
1. Class - Invalid gear in petrol engine
2. file fixed issue
3. gtest, added, test for 3 types of car written

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -5,24 +5,27 @@
...
     std::cout << std::endl << "TOYOTA" << std::endl;
     HybridCar toyota(new PetrolEngine(80, 1400, 5), new ElectricEngine(100, 540));
-    toyota.accelerate(100);
-    toyota.brake();
-    toyota.charge();
-    toyota.refuel();
+    car = &toyota;
+    car->accelerate(100);
+    car->brake();
+    car->restore();*/
```
Comment (WBigaj) [inline main.cpp:30]: Is this PR beta version of Cars considering you have commented main program?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have one question.

File: main.cpp (hunk 1 above)
Comment (majkel84) [reply main.cpp:30]: Comented for clear program Reading, and forget to uncomment it :)

Comment (majkel84) [reply main.cpp]: done

File: PetrolEngine.cpp [hunk 2]
```diff
@@ -10,11 +15,18 @@ PetrolEngine::PetrolEngine(int power, float capacity, int gears)
...
+PetrolEngine::~PetrolEngine() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
 void PetrolEngine::changeGear(int gear)
 {
-    // TODO: Add checking if gear is between -1 and gears_
-    // -1 is for REAR
-    // 0 is for NEUTRAL
+    if (gear < -1) {
```
Comment (MaciejSalamonski) [inline PetrolEngine.cpp:24]: I suggest you tu use constexpr for values like -1

File: ElectricCar.cpp [hunk 3]
```diff
@@ -7,10 +7,23 @@ ElectricCar::ElectricCar(ElectricEngine* engine)
...
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+    delete engine_;
+    std::cout << __FUNCTION__ << std::endl;
+}
+void ElectricCar::restore()       { charge(); }
 void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+void ElectricCar::setEngine(int power, int batteryCapacity) {
+    if (Car::speed_ == 0) {
```
Comment (MaciejSalamonski) [inline ElectricCar.cpp:17]: Maybe it would be better to throw an exception?

File: Car.cpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+Car::~Car()             { std::cout << __FUNCTION__ << std::endl; }
+void Car::turnLeft()    { std::cout << __FUNCTION__ << std::endl; }
+void Car::turnRight()   { std::cout << __FUNCTION__ << std::endl; }
+void Car::brake()       { std::cout << __FUNCTION__ << std::endl; }
+void Car::accelerate(int speed){
+    if (speed >= 0) {
+        speed_ = speed;
+    }
+    std::cout << __FUNCTION__ << speed << std::endl;
```
Comment (MaciejSalamonski) [inline Car.cpp:14]: Ok but what if speed < 0 Maybe it would be better to throw an exception?

**Review by MaciejSalamonski — APPROVED** [review]
Hello, good job! However, I think the tests could be more extensive.

File: PetrolEngine.cpp (hunk 2 above)
Comment (majkel84) [reply PetrolEngine.cpp:24]: done

File: ElectricCar.cpp (hunk 3 above)
Comment (majkel84) [reply ElectricCar.cpp:17]: done

File: Car.cpp (hunk 4 above)
Comment (majkel84) [reply Car.cpp:14]: done

**Review by Kasperice — APPROVED** [review]
Nice code, I did not find anything to change

File: ElectricCar-ut.cpp [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#include <gtest/gtest.h>
+#include "ElectricCar.hpp"
+
+struct ElectricCarTest : ::testing::Test {
+    ElectricCar nissan = ElectricCar(new ElectricEngine(130, 650));
+};
+
+TEST_F(ElectricCarTest, checkAcceleration) {
+    nissan.accelerate(50);
+    ASSERT_EQ(nissan.getSpeed(), 50);
```
Comment (ziobron) [inline ElectricCar-ut.cpp:10]: acceleration to negative values are not covered

File: ElectricCar-ut.cpp [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+TEST_F(ElectricCarTest, checkSetEngine) {
+    ASSERT_EQ(nissan.getEnginePower(), 130);
+    nissan.setEngine(150, 860);
+    ASSERT_EQ(nissan.getEnginePower(), 150);
```
Comment (ziobron) [inline ElectricCar-ut.cpp:16]: 150 is a magic value. You should use constexpr for it.

File: ElectricCar.cpp [hunk 7]
```diff
@@ -1,16 +1,30 @@
...
+        engine_ = new ElectricEngine(power, batteryCapacity);
+        std::cout << "Engine changed" << std::endl;
+    }
+    else {
+        throw std::logic_error("Can't change engine while driving (please verify with Youtube)");
+    }
+}
```
Comment (ziobron) [inline ElectricCar.cpp:25]: nice solution for a problem with providing different kind of engine. However, what if the design forbids it? Eg. we have an EngineFactory class that should produce an engine and we are obliged to get a pointer? It's only hypothetical question ;)

File: ElectricCar.cpp [hunk 8]
```diff
@@ -1,16 +1,30 @@
...
+void ElectricCar::restore()       { charge(); }
 void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+void ElectricCar::setEngine(int power, int batteryCapacity) {
+    if (Car::speed_ == 0) {
+        engine_ = new ElectricEngine(power, batteryCapacity);
+        std::cout << "Engine changed" << std::endl;
+    }
+    else {
+        throw std::logic_error("Can't change engine while driving (please verify with Youtube)");
```
Comment (ziobron) [inline ElectricCar.cpp:23]: 😎

**Review by ziobron — APPROVED** [review]

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
4 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**